### DISCUSSION TITLE
Incrementally add mappings without "invalidating" EntityDescriptorsCache

### DISCUSSION THREAD

**Discussion by boris-turk** [linq2db/linq2db#4269] [Q&A]
Hi,

I have a rather exotic performance issue with linq2db version 5.2.2. I am working on a large ERP system where we have about **150 POCO classes**. If I get it correctly it is best practice to create a single MappingSchema (or a few of them) and fill it with appropriate mappings before executing the first linq2db query. I use fluent API to do that (via `EntityMappingBuilder<>` class).

Here comes the first issue: our method which defines the mappings for a single POCO class takes about **300 milliseconds** to execute for a typical POCO class, which is a bit slow. I profiled a release build with DotTrace. This screenshot clearly shows that most of the time is lost in various linq2db internal methods (e.g., `GetMemberEx`, `GetMappingAttributes`, etc):

![image](https://github.com/linq2db/linq2db/assets/14542258/82f8d932-0a5d-44d4-8c2b-5daadd8dc4e4)

I also realized that the garbage collector is firing up frequently while defining the mappings, which is in a way reasonable (plenty of short lived objects being allocated). This all leads me to the conclusion that I simply cannot speed up significantly the initial mapping configuration. With about 150 POCO classes it looks like we are destined to wait more than **30 seconds** before the mapping configuration is done with its work.

Our ERP system is a desktop application and customers want to have the startup as fast as possible. That's why I tried to parallelize the mapping configuration, i.e., split it to chunks of POCO classes and define on parallel threads the mappings for each chunk.

This leads to my main problem. We use a single MappingSchema through the whole application. It is supposed to hold a cache of all `EntityDescriptor` instances (one EntityDescriptor for one POCO class). I forgot to say before that I also force the building of `EntityDescriptor` instances on a parallel thread during system startup, since the creation of those classes is also slow (reflection, delegate compiling). But as soon as I call the `EntityMappingBuilder<>.Build` method to "flush" the mappings of one POCO class, the `MappingSchema.ConfigurationID` is reset through this call chain:

```
EntityMappingBuilder<>.Build -> FluentMappingBuilder.Build -> MappingSchema.AddMetadataReader -> MappingSchema.ResetID
```

This basically means that all of the `EntityDescriptor` instances built so far and cached in the `EntityDescriptorsCache` structure are no longer accessible (the changed `MappingSchema.ConfigurationID` is part of the cache key).

Long story short: is there a way to incrementally fill the mappings into `MappingSchema` instance while at the same time keeping its `EntityDescriptorsCache` structure intact?

Thanks a lot for any kind of help / suggestions. Best regards,

Boris Turk

**Comment by MaceWindu** [comment]
That's an interesting issue.

Do you really have perfomance issues without entity descriptors pre-generation?

In any case, with your approach I would suggest to try two-phase startup and see how it works for you:

1. configure mappings for entities in parallel
2. prebuild entity descriptors in parallel to avoid cache invalidation

I would like to look into this issue more, so I would appreciate more details on your model:
- how many columns your model has (approx.)
- how many associations your model has (approx.)
- what other mapping objects you configure if any

**Reply by MaceWindu** [reply]
made some tests

150 entities with 12 columns and 2 associations each result in 1 second setup of mapping schema and + 8 seconds to build entity descriptors using single thread

I assume you model is 3-4 times bigger if it takes 30 seconds for you?

**Reply by MaceWindu** [reply]
#4275

**Reply by MaceWindu** [reply]
I will prepare some performance fixes for next release, but you should avoid creation of multiple fluent mapping builders as they introduce major performance issue: each builder on `Build` call produce new instance of metadata reader which results in 150 metadata readers with own attribute caches with a lot of memory pressure. Plus mapping schema needs to interate them all to query for attributes, but this actually has less impact due to caching.

So, to get best performance you should:
1. register entities in single thread using single builder instance (builder is not thread-safe so you cannot parallel it)
2. generate entity descriptors in parallel

**Comment by boris-turk** [comment]
Hi MaceWindu,

let me first apologize for my delayed response to your questions. I've been preoccupied with some other urgent matters. I would also like to thank you for your help, I greatly appreaciate that.

> Do you really have perfomance issues without entity descriptors pre-generation?

Yes, because when certain processes are executed for the first time, they have to do a bunch of distinct queries (with different POCO classes). The whole process takes about 5 seconds and approximately 1.5 second gets consumed by various linq2db methods:
* ``LinqToDB.Mapping.EntityDescriptor.Init``
* ``LinqToDB.Linq.QueryRunner+Mapper`1.GetMapperInfo``
* ``LinqToDB.Mapping.EntityMappingBuilder`1.Property``
* ``LinqToDB.Linq.Builder.ParametersContext.BuildParameter``
* ``LinqToDB.Metadata.MappingAttributesCache.GetMappingAttributes``

I am also adding a dotTrace screenshot which shows that when I exclude the above-mentioned linq2db methods, the consumed time drops by about 1.5s:
![image](https://github.com/linq2db/linq2db/assets/14542258/3e85fbd9-b4cf-4bb5-abad-cde3f659e2c2)

It is true that 1.5s is less than a half of the process which I need to optimize, but I have to start somewhere. I am also not sure if 1.5s all comes from entity descriptor generation or also some other stuff.

Even if I solve all other things and I am left with 1.5 seconds from linq2db, it is still too much. I would like to reduce the reflection and compiling time as much as possible by performing it during application startup phase. I hope you see my point ...

> how many columns your model has (approx.)

There are 124 classes, each class has in average 33 properties.

> how many associations your model has (approx.)

We don't configure linq2db with associations.

> what other mapping objects you configure if any

* [dynamic columns](https://github.com/linq2db/linq2db/issues/744) (determined at runtime), about 15 of them in each table (class)
* about 24 converters for standard clr types, defined with ``MappingSchema.SetConverter<,>``
* 129 enums, each of them approximately 4 members
``FluentMappingBuilder.HasAttribute`` is defined for each member
I also apply converters to each enum:
  * ``MappingSchema.SetConverter<double, TEnum>``
  * ``MappingSchema.SetConverter<TEnum, DataParameter>``
  * ``MappingSchema.SetConverter<TEnum, string>``
  * ``MappingSchema.SetConverter<string, TEnum>``

> you should avoid creation of multiple fluent mapping builders as they introduce major performance issue: each builder on Build call produce new instance of metadata reader which results in 150 metadata readers with own attribute caches with a lot of memory pressure

I didn't think of this one, but it makes sense what you say, thanks for the hint!

> register entities in single thread using single builder instance (builder is not thread-safe so you cannot parallel it)

I actually came up with an interesting solution which seems to work nicely and makes it possible to parallelize also the registration of mappings. I created my own ``IMetadataReader`` implementation which is thread-safe, something like this (for each ``IMetadataReader`` member):

```csharp
MappingAttribute[] IMetadataReader.GetAttributes(Type type)
{
    lock (_lockObject)
    {
        DefineMappingsIfNecessary(type);
        return GetCachedMappings(type);
    }
}
```

I will also test your improvements once they are released. So, I would say that this case is solved. Thanks again for your assistance and best regards!

Boris
